### PR TITLE
Add method CharSequence.occurencesOf

### PR DESCRIPTION
**Pull request by cbruegg**
Related KEEP proposal: [KEEP-20](https://github.com/Kotlin/KEEP/issues/20).

The goal of this pull request is to add an extension method `CharSequence.occurencesOf`, taking two parameters:
1. A pattern to look for in the receiver
2. A flag `ignoreCase`.

It returns a lazy sequence of indices at which the pattern occurs in the receiver.

It is based on the evidentially `O(n)` Knuth-Morris-Pratt algorithm, where `n` is the length of the text to search in. Using `indexOf` repeatedly would have runtime `O(nk)`, where `k` is the length of the pattern.

An abstract use case would be finding all occurences of a short pattern of length `k` in a long text of length `n`, which would be `O(nk)` with `indexOf`, but is `O(n)` with `occurencesOf`. A more concrete use case: Suppose we have a large number of texts (perhaps in a search engine) we would like to run a search query against. With `occurencesOf`, we could get all relevant parts of the text very efficiently, maybe get a substring of the occurence 100 chars before and after, obtaining a summary of where the word occurs. 

This PR of course also includes tests.

### CODE REVIEW

**Comment by cy6erGn0m** [comment]
what about tests for ignore case = true?

File: libraries/stdlib/src/kotlin/text/Strings.kt [hunk 1]
```diff
@@ -1162,3 +1162,77 @@ public fun CharSequence.lineSequence(): Sequence<String> = splitToSequence("\r\n
...
+ * @param [pattern] The pattern to look for in this String. Regular expressions
+ *                 are not supported
+ * @param [ignoreCase] If true, characters are matched even if one is upper and the other is
+ *                 lower case
+ *
+ * @return A list of indices where the supplied [pattern] starts in the text.
+ */
+public fun CharSequence.occurrencesOf(pattern: CharSequence, ignoreCase: Boolean = false): Sequence<Int> {
+
+    if (isEmpty() || pattern.isEmpty()) {
```
Comment (cy6erGn0m) [inline libraries/stdlib/src/kotlin/text/Strings.kt]: `if (pattern.isEmpty() || pattern.length > this.length)`  ?

**Comment by cbruegg** [comment]
There's this one: `assertEquals("Abaabaaababac".occurrencesOf("abaa", ignoreCase = true).toList(), listOf(0, 3))`

I'll add some more though. Good suggestion for the early return as well, I'll change it and update this PR. Thanks for the feedback!

**Comment by cy6erGn0m** [comment]
I also believe we need few tests for unicode characters

**Comment by cy6erGn0m** [comment]
Another one could be `assertEquals("abaabaaababac".occurrencesOf("abAa", ignoreCase = true).toList(), listOf(0, 3))`

**Comment by cbruegg** [comment]
I've just added tests for two greek unicode characters and added more tests for the ignoreCase condition.

File: libraries/stdlib/src/kotlin/text/Strings.kt [hunk 2]
```diff
@@ -1162,3 +1162,77 @@ public fun CharSequence.lineSequence(): Sequence<String> = splitToSequence("\r\n
...
+ *                 are not supported
+ * @param [ignoreCase] If true, characters are matched even if one is upper and the other is
+ *                 lower case
+ *
+ * @return A list of indices where the supplied [pattern] starts in the text.
+ */
+public fun CharSequence.occurrencesOf(pattern: CharSequence, ignoreCase: Boolean = false): Sequence<Int> {
+
+    if (pattern.isEmpty() || pattern.length > this.length) {
+        return emptySequence()
```
Comment (ilya-g) [inline libraries/stdlib/src/kotlin/text/Strings.kt]: I believe this should be consistent with the result of `Regex("").findAll(input = "....")`, i.e. a match in every position.

**Comment by cbruegg** [comment]
What do you mean? If you take a look at the tests, `"abaabaaababac".occurrencesOf("abaa")` returns 0 and 3, so it does match every position. I haven't run the regex you've posted, but it should have the same result.

**Comment by cbruegg** [comment]
Nevermind, I see what you mean. I'll change the PR.

**Comment by cy6erGn0m** [comment]
Should we return *all* occurrences or only non-overlapping? eg
`AAA.occurrencesOf(AA)` = `0, 1` or `0` ?

**Comment by cbruegg** [comment]
I could probably add an additional parameter for that. Its default value
should be to find all occurrences in my opinion.

Sergey Mashkov <notifications@github.com> schrieb am Mi., 17. Feb. 2016
11:47:

**Comment by cy6erGn0m** [comment]
@ilya-g what do you think? what behaviour is better by default? as far as I understand all our api doesn't return overlapping occurrences, isn't it?

**Comment by cbruegg** [comment]
The equivalent regex does not return overlapping occurrences. I'll set the
default parameter value to non-overlapping.

Sergey Mashkov <notifications@github.com> schrieb am Mi., 17. Feb. 2016
12:35:

**Comment by cbruegg** [comment]
Just checking you've noticed the commit - not sure if GitHub sends notifications for commits in PRs or only for comments. If you saw the commit already, please disregard this comment.

File: libraries/stdlib/src/kotlin/text/Strings.kt [hunk 3]
```diff
@@ -1162,3 +1162,84 @@ public fun CharSequence.lineSequence(): Sequence<String> = splitToSequence("\r\n
...
+ * @return A list of indices where the supplied [pattern] starts in the text.
+ */
+public fun CharSequence.occurrencesOf(pattern: CharSequence, ignoreCase: Boolean = false, matchOverlapping: Boolean = false): Sequence<Int> {
+
+    if (pattern.length > this.length) {
+        return emptySequence()
+    }
+
+    if (pattern.isEmpty()) {
+        return indices.asSequence()
```
Comment (ilya-g) [inline libraries/stdlib/src/kotlin/text/Strings.kt]: It should be `(0..length).asSequence()`, because you need to count an empty match before the first character and empty one after the last characater.

File: libraries/stdlib/src/kotlin/text/Strings.kt [hunk 4]
```diff
@@ -1162,3 +1162,84 @@ public fun CharSequence.lineSequence(): Sequence<String> = splitToSequence("\r\n
...
+    }
+
+    if (pattern.length == 1) {
+        return indices.asSequence().filter { this[it].equals(pattern[0], ignoreCase) }
```
Comment (ilya-g) [inline libraries/stdlib/src/kotlin/text/Strings.kt]: We could extract `CharSequence.occurrencesOf(char: Char, ignoreCase)` function from here.

**Comment by cbruegg** [comment]
Thanks for the suggestions, I've incorporated them.

File: libraries/stdlib/src/kotlin/text/Strings.kt [hunk 5]
```diff
@@ -1162,3 +1162,97 @@ public fun CharSequence.lineSequence(): Sequence<String> = splitToSequence("\r\n
  * * Splits this char sequence to a list of lines delimited by any of the following character sequences: CRLF, LF or CR.
  */
 public fun CharSequence.lines(): List<String> = lineSequence().toList()
+
+/**
+ * For each i in [0, length), this function computes
+ * the length of the longest suffix of a substring of pattern from 0 to i
+ * that is also a prefix of the pattern itself.
+ */
+private fun computePrefixFunction(pattern: CharSequence): IntArray {
```
Comment (ilya-g) [inline libraries/stdlib/src/kotlin/text/Strings.kt]: I'm not sure the prefix function is computed correctly. For the following example "ABCDABD" it returns [0,0,0,0,1,2,0], while according to the [reference](https://en.wikipedia.org/wiki/Knuth%E2%80%93Morris%E2%80%93Pratt_algorithm#Worked_example_of_the_table-building_algorithm) it should be [-1,0,0,0,0,1,2].

Also it loops infinitely on the input "AAABAAB".

File: libraries/stdlib/test/text/StringTest.kt [hunk 6]
```diff
@@ -714,6 +714,43 @@ class StringTest {
         }
     }
 
+    @test fun occurencesOf() {
+        // Multiple overlapping occurances
+        assertEquals("abaabaaababac".occurrencesOf("abaa", matchOverlapping = true).toList(), listOf(0, 3))
```
Comment (ilya-g) [inline libraries/stdlib/test/text/StringTest.kt]: `expected` value comes first, `actual` value second.

File: libraries/stdlib/src/kotlin/text/Strings.kt (hunk 5 above)
Comment (cbruegg) [inline libraries/stdlib/src/kotlin/text/Strings.kt]: I've fixed the infinite loop, the prefix function algorithm missed a subtraction that wasn't caught by a test yet, so I've also added a test case for that.

I've never seen -1 in the table before, so KMP as described in the Wikipedia article seems to be a slightly different variation of KMP. My implementation is based on the pseudocode we were given in a lecture on Fundamental Algorithms at the University of Paderborn and it was also proven to be working correct by the lecturer, so this should be fine.

File: libraries/stdlib/test/text/StringTest.kt (hunk 6 above)
Comment (cbruegg) [inline libraries/stdlib/test/text/StringTest.kt]: Fixed, thanks.

File: libraries/stdlib/src/kotlin/text/Strings.kt (hunk 5 above)
Comment (ilya-g) [inline libraries/stdlib/src/kotlin/text/Strings.kt]: Should the `ignoreCase` parameter affect prefix function?

File: libraries/stdlib/src/kotlin/text/Strings.kt [hunk 7]
```diff
@@ -1162,3 +1162,97 @@ public fun CharSequence.lineSequence(): Sequence<String> = splitToSequence("\r\n
...
+    var matches = 0
+    return generateSequence {
+        while (i < length) {
+            while (matches > 0 && !pattern[matches].equals(this[i], ignoreCase)) {
+                matches = prefixFunction[matches - 1]
+            }
+
+            if (pattern[matches].equals(this[i], ignoreCase)) {
+                matches++
+            }
```
Comment (ilya-g) [inline libraries/stdlib/src/kotlin/text/Strings.kt]: This approach requires two character comparisons per position when `matches > 0`. Could you restructure the loop body? Also applies to `computePrefixFunction`.

File: libraries/stdlib/src/kotlin/text/Strings.kt (hunk 5 above)
Comment (cbruegg) [inline libraries/stdlib/src/kotlin/text/Strings.kt]: Actually yes, it should. Learning the hard way that even with tests such things can easily be missed, sorry. I've just pushed a fix with another test case as well.

File: libraries/stdlib/src/kotlin/text/Strings.kt (hunk 7 above)
Comment (cbruegg) [inline libraries/stdlib/src/kotlin/text/Strings.kt]: Done.

**Comment by ilya-g** [comment]
Since we have no plans merging this PR in the standard library (see [this comment](https://github.com/Kotlin/KEEP/issues/20#issuecomment-230317194) in the related KEEP discussion), I'm closing this for now.

Since it's linked to the KEEP, it would be easy to recover it when it will become required.
